content/en/blog/2023-02-07-building-a-raft-part-3.md

```diff
@@ -80,25 +80,28 @@ Here's the full logic for the `RequestVote` function:
 func (s *RaftServer) RequestVote(ctx context.Context, in *api.VoteRequest) (out *api.VoteReply, err error) {
 	s.Lock()
 	defer s.Unlock()
-	out = &api.VoteReply{Id: s.id, Term: s.currentTerm}
+	out = &api.VoteReply{Id: s.id, Term: s.currentTerm, VoteGranted: false}
 	lastLogIndex, lastLogTerm := s.lastLogIndexAndTerm()
 
+	// Check if the incoming request has a higher term than ours, if so then there is a node in the cluster 
+	// more up to date, so we ensure we are a follower and reset our election timeout
 	if in.Term > s.currentTerm {
 		fmt.Printf("RequestVote: in.Term > s.currentTerm, reverting to follower\n")
 		s.becomeFollower(in.Term)
 	}
 
+	// This (very complicated) check is to make sure the following things are true before granting the vote:
+	//	1. This nodes current term is the same as the requester's term
+	//  2. This node hasn't voted for another candidate
+	//  3. The lastLogTerm and LastLogIndex of the requestor is at least up to date with this node's
 	if s.currentTerm == in.Term && (s.votedFor == in.CandidateId || s.votedFor == "") &&
 		(int(in.LastLogTerm) > lastLogTerm || (in.LastLogTerm == int32(lastLogTerm) &&
 			in.LastLogIndex >= int64(lastLogIndex))) {
 		fmt.Printf("granting vote to %v\n", in.CandidateId)
 		out.VoteGranted = true
 		s.votedFor = in.CandidateId
 		s.lastHeartbeat = time.Now()
-	} else {
-		out.VoteGranted = false
-	}
-	out.Term = s.currentTerm
```

Comment:
Can I ask why this got deleted?

Reply:
Based on a previous comment of yours I refactored the code so that `VoteGranted` is made explicitly false when I instantiate the `VoteReply`. While doing that I noticed that since `s` is locked, `s.currentTerm` won't change. Since that is also set on the  `VoteReply` instantiation it doesn't need to be set again. I just thought it made the code cleaner and easier for readers to understand.

Reply:
Gotcha - my comment from before wasn't meant to criticize what you had written previously, but rather to prompt you to explain why it is **critical** that the receiving server sends its current term back to the candidate seeking election. I think we still need to explain this.

Reply:
Updated :thumbsup: